src/librustc_typeck/astconv.rs

```diff
@@ -229,17 +233,26 @@ fn report_elision_failure(
             m.push_str(", ");
         }
     }
-    if len == 1 {
-        fileline_help!(tcx.sess, default_span,
-                       "this function's return type contains a borrowed value, but \
-                        the signature does not say which {} it is borrowed from",
-                       m);
-    } else if len == 0 {
+
+    if len == 0 {
         fileline_help!(tcx.sess, default_span,
                        "this function's return type contains a borrowed value, but \
                         there is no value for it to be borrowed from");
         fileline_help!(tcx.sess, default_span,
                        "consider giving it a 'static lifetime");
+    } else if !any_lifetimes {
+        fileline_help!(tcx.sess, default_span,
```

Comment:
I was thinking the previous message implied that there were no arguments to draw from, and I was hoping to make the message as explicit as possible; I think if they were made the same they would be slightly more confusing. I don't feel strongly on this or anything, I'd be happy to change it so they have the same message if that's the consensus.

Also yeah, I'll do the tests.